src/instance/device/file.rs

```diff
@@ -0,0 +1,70 @@
+use async_trait::async_trait;
+use chrono::Utc;
+use std::time;
+use tokio::{fs::File, io::AsyncWriteExt};
+
+use crate::models;
+
+use super::{common::*, DeviceError};
+
+pub type FileDevice = Rewriter<FileDeviceImpl>;
+
+pub struct FileDeviceImpl {
+    leds: Vec<models::Color>,
+    print_timestamp: bool,
+    file_handle: File,
+    last_write_time: time::Instant,
+}
+
+#[async_trait]
+impl WritingDevice for FileDeviceImpl {
+    type Config = models::File;
+
+    fn new(config: &Self::Config) -> Result<Self, DeviceError> {
+        let file_handle = std::fs::OpenOptions::new()
+            .write(true)
+            .create(true)
+            .append(true)
+            .open(&config.output)?;
+
+        Ok(Self {
+            leds: vec![Default::default(); config.hardware_led_count as _],
+            print_timestamp: config.print_time_stamp,
+            file_handle: File::from_std(file_handle),
+            last_write_time: time::Instant::now(),
+        })
+    }
+
+    async fn set_let_data(
+        &mut self,
+        _config: &Self::Config,
+        led_data: &[models::Color],
+    ) -> Result<(), DeviceError> {
+        self.leds.copy_from_slice(led_data);
+        Ok(())
+    }
+
+    async fn write(&mut self) -> Result<(), DeviceError> {
+        let mut strbuf = String::new();
```

Comment:
Is there a specific reason for buffering to a string before writing? Since we're calling flush later on file_handle, it might be better to use a [BufWriter](https://docs.rs/tokio/latest/tokio/io/struct.BufWriter.html) to handle the buffering for us.

Reply:
If we do want to handle buffering ourselves, it'll be better to move the strbuf to self to reuse its allocation, and call `String::clear()` on every write.

Reply:
I used buffering into a string so its nicer syntax-wise when constructing the strings..

For a BufWriter it would look like this

```rs
let mut buf_writer = BufWriter::new(...);

...

buf_writer.write(format!("Some fmt {}", "bla").as_bytes()).await?;
```

I would prefer `write!(str_buf, "Some fmt {}", "bla");`...

Reply:
I thought `tokio::fs::File` had a `write_fmt` method but it doesn't seem to be the case, so you're right we need the intermediary `String` anyways!